src/net/server/handlers/mul.rs

```diff
@@ -0,0 +1,439 @@
+
+#[async_trait]
+impl<B: Send> FromRequest<B> for Path {
+    type Rejection = MpcServerError;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let extract::Path((query_id, step)) =
+            extract::Path::<(QueryId, UniqueStepId)>::from_request(req).await?;
+        Ok(Path(query_id, step))
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", derive(serde::Deserialize))]
+pub struct IdentityQuery {
+    identity: Identity,
+}
+
+/// After an [`OwnedPermit`] has been reserved, it can be used once to send an item on the channel.
+/// Panics if cloned while a permit exists.
+pub struct ReservedPermit<T>(Option<mpsc::OwnedPermit<T>>);
+
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn new(permit: mpsc::OwnedPermit<T>) -> Self {
+        Self(Some(permit))
+    }
+    /// # Panics
+    /// if called more than once
+    pub fn send(&mut self, item: T) {
+        self.0
+            .take()
+            .expect("should only call `send` once")
+            .send(item);
+    }
+}
+
+impl<T> Clone for ReservedPermit<T> {
+    /// # Panics
+    /// if a permit exists
+    fn clone(&self) -> Self {
+        assert!(self.0.is_none());
+        Self(None)
+    }
+}
+
+pub async fn obtain_permit_mw<T: Send + 'static, B>(
+    sender: mpsc::Sender<T>,
+    mut req: Request<B>,
+    next: Next<B>,
+) -> Result<Response, MpcServerError> {
+    let permit = sender.reserve_owned().await?;
+    req.extensions_mut().insert(ReservedPermit::new(permit));
+    Ok(next.run(req).await)
+}
+
+/// extracts the [`MessageChunks`] from the request and forwards it to the Message layer via the
+/// `permit`. If we try to extract the `permit` via the `Extension`'s `FromRequest` implementation,
+/// it will call `.clone()` on it, which will remove the `OwnedPermit`. Thus, we must access the
+/// `permit` via `Request::extensions_mut`, which returns [`Extensions`] without cloning.
+#[allow(clippy::unused_async)] // handler is expected to be async
+#[allow(clippy::cast_possible_truncation)] // length of envelopes array known to be less u32
+pub async fn handler(
+    Path(_query_id, step): Path,
+    // TODO: we shouldn't trust the client to tell us their identity.
+    //       revisit when we have figured out discovery/handshake
+    Query(IdentityQuery { identity }): Query<IdentityQuery>,
+    headers: RecordHeaders,
+    mut req: Request<Body>,
+) -> Result<(), MpcServerError> {
+    // prepare data
+    let channel_id = ChannelId { identity, step };
+
+    let body = hyper::body::to_bytes(req.body_mut()).await?;
+    headers.matches_body(body.len())?;
+    let envelopes = body
+        .as_ref()
+        .chunks(headers.data_size as usize)
+        .enumerate()
+        .map(|(record_id, chunk)| MessageEnvelope {
+            record_id: RecordId::from(headers.offset + record_id as u32),
+            payload: chunk.to_vec().into_boxed_slice(),
+        })
+        .collect::<Vec<_>>();
```

Comment:
I'm welcome to an alternative suggestion. What would we send to messaging layer if it's not a vec (which is what is [defined in `MessageChunks` now](https://github.com/private-attribution/ipa/blob/cea96f47a0d3d74e1bf4efdf03e190b70097efe9/src/helpers/fabric.rs#L21). The `.send` necessitates `Send + 'static`

Reply:
Hmm, yeah, that is awkward.  Perhaps we should track this in an issue.

@akoshelev do you think that an interface to the network that is more along the lines of `send(&self, channel: ChannelId) -> Stream<Message>` would be acceptable?  (The same for receiving.)  That might be less work for Thurston to manage even as many networking libraries try hard to make stream-level interactions independent of each other.

Reply:
yea that was the alternative I considered when working on #130. I thought it might be easier for us to implement it this way, but I am not so sure right now. The thing that tripped me off was the fact that infra cannot pull streams it never received requests to receive in this case. I also wasn't sure whether I can make it efficient by letting infra to guess which channel is ready. We could probably mitigate that by adding another method to `Network` that can hint which stream has the stream of bytes ready. 

any thoughts?